### PR TITLE
Fix CLI context server precedence

### PR DESCRIPTION
**Pull request by srstack**
## Summary
- make active context server override DRIVE9_SERVER during credential resolution
- honor explicit `<ctx>:/path` contexts for CLI fs commands that accept remote path prefixes (`ls`, `cp`, `mv`, `rm`, `mkdir`, `cat`, `grep`, `stat`, `chmod`, `find`)
- reject mixed explicit/current-context server-side copy and rename operations that could target the wrong tenant
- add regression coverage for remote-context selection and server precedence

## Testing
- go test ./cmd/drive9/cli
- go build -o bin/drive9 ./cmd/drive9

## Summary by CodeRabbit

* **New Features**
  * Unified remote-context resolution applied across filesystem commands with consistent handling and early error returns.
  * Explicit rejection of cross-context operations with clearer error messages.

* **Bug Fixes**
  * Credential/server resolution precedence refined: active-context server now beats DRIVE9_SERVER; env still overrides config when context server is absent.

* **Tests**
  * Expanded tests for explicit/delegated remote contexts and cross-context command behavior.

* **Documentation**
  * Updated docs/specs and quickstart guidance to reflect new server-selection precedence and migration notes.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mem9-ai/drive9/pull/454?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: cmd/drive9/cli/client.go [hunk 1]
```diff
@@ -58,3 +60,44 @@ func NewFromEnvWithWarm() *client.Client {
...
+}
+
+func fsClientForRemoteArg(defaultClient *client.Client, raw string) (*client.Client, string, string, bool, error) {
+	rp, isRemote := ParseRemote(raw)
+	if !isRemote {
+		return defaultClient, raw, "", false, nil
+	}
+	if rp.Context == "" {
+		return defaultClient, rp.Path, "", true, nil
+	}
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/client.go:103]: Addressed by narrowing the PR scope rather than changing `fs sh` in this PR. The PR description now explicitly covers only `fs` commands that accept remote path prefixes directly (`ls`, `cp`, `mv`, `rm`, `mkdir`, `cat`, `grep`, `stat`, `chmod`, `find`), and no longer claims `fs sh` support.

File: cmd/drive9/cli/cp.go [hunk 2]
```diff
@@ -111,6 +111,22 @@ func Cp(c *client.Client, args []string) error {
...
+		return fmt.Errorf("cross-context copy not supported: %s -> %s", srcRP.Context, dstRP.Context)
+	case srcIsRemote && srcRP.Context != "":
+		var err error
+		c, err = newFSClientForContext(srcRP.Context)
+		if err != nil {
+			return err
+		}
+	case dstIsRemote && dstRP.Context != "":
+		var err error
+		c, err = newFSClientForContext(dstRP.Context)
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/cp.go]: Fixed in 63241da. I removed the early client rebinding and now reject mixed explicit/current-context remote-to-remote copies instead of silently rebinding `:/...` to the other side. That preserves `:/` semantics and prevents cross-tenant writes in cases like `prod:/a -> :/b`.

File: cmd/drive9/cli/client.go [hunk 3]
```diff
@@ -58,3 +60,44 @@ func NewFromEnvWithWarm() *client.Client {
...
+func newFSClientForContext(name string) (*client.Client, error) {
+	cfg := loadConfig()
+	ctx := cfg.Contexts[name]
+	if ctx == nil {
+		return nil, fmt.Errorf("context %q not found; run: drive9 ctx ls", name)
+	}
+
+	server := ctx.Server
+	if server == "" {
+		server = cfg.ResolveServer()
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/client.go]: Fixed in 63241da. `newFSClientForContext` no longer falls back through `cfg.ResolveServer()`; when the named context has no `server`, it now falls back to `cfg.Server` and then the compiled default, so explicit `ctx:/...` lookups are no longer coupled to the active context.

[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/client.go]: Fixed in 63241da. Explicit context lookup no longer uses the active-context-aware `cfg.ResolveServer()` fallback. When a named context omits `server`, the fallback is now `cfg.Server` and then the compiled default, which keeps explicit `ctx:/...` routing independent from the active context.

File: cmd/drive9/cli/mv.go [hunk 4]
```diff
@@ -16,12 +16,25 @@ func Mv(c *client.Client, args []string) error {
...
+	newClient := c
+	newClient, newPath, newCtx, _, err = fsClientForRemoteArg(newClient, newPath)
+	if err != nil {
+		return err
+	}
+	if oldCtx != "" && newCtx != "" && oldCtx != newCtx {
+		return fmt.Errorf("cross-context rename not supported: %s -> %s", oldCtx, newCtx)
+	}
+	if oldCtx == "" && newCtx != "" {
+		c = newClient
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/mv.go]: Fixed in 63241da. `mv` now keeps `:/...` bound to the current context and rejects mixed explicit/current-context renames instead of rebinding both operands to the explicit side. That avoids renaming the wrong tenant object for inputs like `:/old -> prod:/new`.

File: cmd/drive9/cli/credentials_test.go [hunk 5]
```diff
@@ -196,13 +196,34 @@ func TestResolveCredentials_ServerEnvOverridesContextServer(t *testing.T) {
...
 	}
 	// Credential still from config.
 	if r.APIKey != "sk-config-owner" {
 		t.Fatalf("APIKey = %q, want config", r.APIKey)
 	}
+	if r.ServerSource != "config:prod" {
+		t.Fatalf("ServerSource = %q", r.ServerSource)
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/credentials_test.go:210]: Fixed in 9355e3c. Server resolution is now independent from credential selection: the active context's Server is loaded first and wins even when the credential itself comes from DRIVE9_API_KEY or DRIVE9_VAULT_TOKEN. I also added explicit regression coverage for both env-owner-key and env-vault-token cases in credentials_test.go.

File: cmd/drive9/cli/cp.go [hunk 6]
```diff
@@ -152,6 +152,13 @@ func Cp(c *client.Client, args []string) error {
 
 	switch {
 	case src == "-" && dstIsRemote:
+		if dstRP.Context != "" {
+			var err error
+			c, err = newFSClientForContext(dstRP.Context)
+			if err != nil {
+				return err
+			}
+		}
```
[the remark this answers is not in the document]
Comment (srstack) [reply cmd/drive9/cli/cp.go:186]: Fixed in 9355e3c. The recursive cp path now applies the same explicit-context client routing as the single-file path: local->remote and remote->local resolve the named context client before calling copyTree*, and remote->remote rejects mixed explicit/current-context copies while allowing same-context explicit copies. I also added a recursive mixed-context regression test in fs_context_test.go.

**Review by qiffang — CHANGES_REQUESTED** [review]
## PR #454 Review — REQUEST_CHANGES

### B1 (BLOCKER): Server precedence change contradicts documented spec

The PR changes server resolution from:
```
DRIVE9_SERVER > active-context.server > config.server > default
```
to:
```
active-context.server > DRIVE9_SERVER > config.server > default
```

This directly contradicts two existing spec documents:

1. **`docs/specs/vault-interaction-end-state.md` L490:**
   > `ctx use` does **not** lock server and credential together: if `DRIVE9_SERVER` is set, it overrides the context's `server` field even when the active context is used for credentials.

2. **`docs/guides/vault-quickstart.md` L261:**
   > `DRIVE9_SERVER` is resolved independently: if set, it overrides the active context's `server` field even when credentials come from the active context.

If the precedence change is intentional, the spec docs **must** be updated in the same PR. Otherwise users relying on `DRIVE9_SERVER` to override context server (e.g. CI scripts, Docker containers with env vars) will silently hit the wrong server after upgrading.

**Question for @qiffang**: Is the intent to make context server always win over env? If yes, this is a deliberate spec change and needs doc updates + a migration note. If no, the credentials.go change should be reverted.

### B2: `newFSClientForContext` server fallback uses `cfg.Server` but not `DRIVE9_SERVER`

```go
func newFSClientForContext(name string) (*client.Client, error) {
    ...
    server := ctx.Server
    if server == "" {
        if cfg.Server != "" {
            server = cfg.Server
        } else {
            server = defaultServerURL
        }
    }
    ...
}
```

When using explicit `fork:/path` syntax, if the `fork` context has no `server` field, it falls back to `cfg.Server` → `defaultServerURL`. But it skips `DRIVE9_SERVER` env entirely. This creates an inconsistency: for the *active* context, `DRIVE9_SERVER` is checked (in the new code, as a fallback); for *explicit* contexts, it's not.

Either `DRIVE9_SERVER` should be checked here too, or the asymmetry needs to be documented with rationale.

### B3: `fsClientForRemoteArg` returns 5 values but callers ignore most

```go
c, path, _, _, err = fsClientForRemoteArg(c, path)
```

Every single-path command (ls, cat, stat, chmod, find, grep, rm, mkdir) ignores the context name and isRemote bool. The function signature is designed for cp/mv's cross-context check, but pollutes simpler call sites. Consider splitting into two helpers or using a struct return.

This is a style nit, not a blocker — but the unused return values make the diff noisier.

### ✅ Good parts

1. Cross-context copy/rename rejection is clean and correct — both recursive and non-recursive paths are covered
2. Test coverage is comprehensive: explicit context ls, delegated rejection, cross-context cp/mv rejection, config server fallback
3. `newFSClientForContext` correctly rejects delegated contexts for fs commands
4. Uniform `fsClientForRemoteArg` replaces scattered `ParseRemote` calls — good DRY improvement

### Summary

B1 is a blocker: the server precedence change contradicts documented spec. Either update the spec docs in this PR or revert the precedence change. B2 should be addressed (env fallback consistency). B3 is optional.

**Review by qiffang — COMMENTED** [review]
## adversary-2 review — PR #454 Fix CLI context server precedence

### Scope
16 files changed. Two semantic changes:
1. **Server precedence flip**: `active-context.server > DRIVE9_SERVER > config.server > default` (was: `DRIVE9_SERVER > active-context.server > ...`)
2. **Explicit remote-context dispatch**: `<ctx>:/path` arguments now resolve their own client via `newFSClientForContext`, applied to ls/cat/stat/rm/mkdir/chmod/grep/find/cp/mv

---

### B1 — BLOCKER: `cp` cross-context check has a silent pass-through for `(explicit, bare-remote)` pairs

In `cp.go`, the cross-context guard for both recursive and non-recursive remote→remote copies is:

```go
switch {
case srcRP.Context == "" && dstRP.Context == "":
case srcRP.Context != "" && dstRP.Context != "" && srcRP.Context == dstRP.Context:
    // ...
default:
    return fmt.Errorf("cross-context copy not supported: ...")
}
```

This correctly rejects `prod:/a → staging:/b`. But consider: `prod:/a → :/b` (one explicit context, one bare remote). `ParseRemote(":/b")` returns `{Context:"", Path:"/b"}`, so `srcRP.Context="prod"` and `dstRP.Context=""`. This hits the `default` branch → rejected.

However, the **reverse** case `:/a → prod:/b` (bare src, explicit dst) also lands in `default` → rejected, which is correct.

But what about `prod:/a → /b`? `ParseRemote("/b")` returns `(_, false)`, so `dstIsRemote=false`, and we take the `srcIsRemote && !dstIsRemote` branch (remote→local download) — which is fine.

And `prod:/a → -` (stdout)? Also fine, separate branch.

OK — the cp cross-context guard is actually correct for all combinations. **Withdrawing B1.**

### B2 — BLOCKER: `mv.go` cross-context check doesn't use `isRemote` flag; bare local path with colon in name would be misclassified

In `mv.go`:

```go
oldRP, oldIsRemote := ParseRemote(oldPath)
newRP, newIsRemote := ParseRemote(newPath)
if oldIsRemote {
    oldPath = oldRP.Path
}
if newIsRemote {
    newPath = newRP.Path
}

switch {
case oldRP.Context == "" && newRP.Context == "":
```

The switch checks `oldRP.Context` and `newRP.Context` without gating on `oldIsRemote`/`newIsRemote`. If **neither** argument is remote (`oldIsRemote=false, newIsRemote=false`), `ParseRemote` returns `RemotePath{}` for both, so `oldRP.Context=""` and `newRP.Context=""` → first case matches → falls through. This works.

But the `oldIsRemote`/`newIsRemote` variables are computed and never used elsewhere. That's dead code after this switch. Not a correctness bug, but the two unused variables should be cleaned up (Go will allow them since they're technically used in the `if` guards above, but semantically the cross-context switch is not using them).

**Downgrading to warning W1**: unused semantic intent — `oldIsRemote`/`newIsRemote` should inform the cross-context switch.

### B3 — BLOCKER: `newFSClientForContext` calls `loadConfig()` on every invocation — double config reads per command

`fsClientForRemoteArg` → `newFSClientForContext` → `loadConfig()`. But the caller already has a client built from `NewFromEnv()` → `ResolveCredentials()` → `loadConfig()` (cached via `sync.Once`). The second `loadConfig()` in `newFSClientForContext` is not cached — it re-reads `~/.drive9/config` from disk every time.

For a single `ls fork:/dir` this is one extra file read. For `cp -r` with many files, it's one per call site in cp.go (up to 6 branches each calling `newFSClientForContext`). This is not a correctness bug per se, but it's an **inconsistency hazard**: if the config file changes between the two reads (e.g. concurrent `ctx use` in another terminal), the credential resolver and the FS client could target different servers.

**BLOCKER**: `newFSClientForContext` should either accept a `*Config` parameter or cache the config read, not silently re-read from disk.

### B4 — BLOCKER: `newFSClientForContext` server fallback chain skips `DRIVE9_SERVER` env var

`newFSClientForContext` has:
```go
server := ctx.Server
if server == "" {
    if cfg.Server != "" {
        server = cfg.Server
    } else {
        server = defaultServerURL
    }
}
```

The new precedence for the unified resolver (`resolveCredentialsWithConfig`) is:
`active-context.server > DRIVE9_SERVER > config.server > default`

But `newFSClientForContext` skips `DRIVE9_SERVER` entirely — it goes `ctx.Server > config.server > default`. This means `drive9 ls fork:/dir` with `DRIVE9_SERVER=https://override.example` and `fork` context having no server field will use `config.server` instead of `DRIVE9_SERVER`.

This is an inconsistent precedence contract between the two resolution paths. The explicit-context path and the active-context path should agree on server fallback order.

**BLOCKER**: Either document why the explicit-context path intentionally differs, or align the fallback chain to match `resolveCredentialsWithConfig`.

### W2 — WARNING: No test for `newFSClientForContext` server fallback to `defaultServerURL`

Tests cover explicit context server and config.server fallback, but there's no test for the case where both `ctx.Server` and `cfg.Server` are empty and the function falls through to `defaultServerURL`.

### W3 — WARNING: `fsClientForRemoteArg` returns 5 values; 3rd and 4th are unused at every call site

```go
func fsClientForRemoteArg(...) (*client.Client, string, string, bool, error)
```

The `string` (context name) and `bool` (isRemote) returns are never used by any caller — all call sites use `_, _` for positions 3 and 4. Dead return values add API surface without benefit.

### W4 — WARNING: `cp.go` has 6 near-identical `newFSClientForContext` call blocks

Each branch in `Cp` that handles an explicit context repeats:
```go
if XRP.Context != "" {
    var err error
    c, err = newFSClientForContext(XRP.Context)
    if err != nil {
        return err
    }
}
```

This is duplicated 6 times. One early-resolution pass before the switch would eliminate the duplication and reduce error surface.

---

### Checklist

| Gate | Status |
|------|--------|
| All inputs handled? | ⚠️ B4: DRIVE9_SERVER skipped in explicit-context path |
| Edge cases covered? | ⚠️ W2: no defaultServerURL fallback test |
| Failure behavior defined? | ✅ Clear errors for delegated/missing/unknown types |
| API explicit? | ⚠️ W3: dead return values; W1: dead variables |
| Race/state issues? | ⚠️ B3: double config read creates TOCTOU window |
| Regression tests? | ✅ 6 new tests covering key scenarios |

### Verdict: **REVISE**

**Must fix (blockers)**:
- **B3**: `newFSClientForContext` re-reads config from disk — either accept `*Config` param or use a shared cached load
- **B4**: Explicit-context server fallback skips `DRIVE9_SERVER` — inconsistent with unified resolver

**Should fix (warnings)**:
- W1: Remove or use `oldIsRemote`/`newIsRemote` in mv.go cross-context switch
- W2: Add test for defaultServerURL fallback in explicit-context path
- W3: Remove unused return values from `fsClientForRemoteArg` or justify them
- W4: Deduplicate 6 identical `newFSClientForContext` blocks in cp.go

**Review by qiffang — CHANGES_REQUESTED** [review]
REQUEST_CHANGES from @dat9-dev2.

I verified the diff at `9355e3c`, read the affected CLI credential/context code, checked the current docs/specs, and ran:

- `/usr/local/go/bin/go test ./cmd/drive9/cli -count=1`
- `/usr/local/go/bin/go build -o /tmp/drive9-pr454-drive9 ./cmd/drive9`

Targeted tests/build pass, but I do not think this PR is safe to merge yet.

## Blockers

### B1 — Server precedence change contradicts the locked spec/docs

The PR changes `ResolveCredentials()` server precedence to:

```text
active-context.server > DRIVE9_SERVER > config.server > default
```

Current normative docs say the opposite:

- `docs/specs/vault-interaction-end-state.md` §14.2 lines 484-490: `DRIVE9_SERVER` overrides active context server.
- `docs/guides/vault-quickstart.md` line 261: `DRIVE9_SERVER` overrides active context `server` even when credentials come from the active context.

This is not just a comment drift. It changes CI/Docker/one-off shell behavior where operators intentionally set `DRIVE9_SERVER` as an orthogonal routing override. With this PR, an active context with `server` silently ignores `DRIVE9_SERVER` and may route credentials to a different tenant/server than the caller requested.

Required fix: either restore the documented precedence (`DRIVE9_SERVER > active-context.server`) or explicitly update the spec/guide and add migration notes/tests for the breaking behavior change.

### B2 — Explicit `<ctx>:/path` server fallback skips `DRIVE9_SERVER`

`newFSClientForContext()` resolves server as:

```text
ctx.server > config.server > default
```

It never considers `DRIVE9_SERVER`. That makes explicit context routing inconsistent with the unified resolver and the documented server-override model. Example: `DRIVE9_SERVER=https://staging drive9 fs ls fork:/` with `fork` containing no server should route to staging if env server remains an orthogonal override; this PR routes to config/default instead.

Required fix: make explicit context server fallback follow the same documented server precedence, or update the spec if explicit contexts intentionally ignore env routing. If ignoring env is intended, that exception needs to be documented and tested because it is a surprising split from `ResolveCredentials()`.

### B3 — `newFSClientForContext()` reloads config independently of the cached resolver snapshot

`ResolveCredentials()` uses `sync.Once` so a CLI invocation has a stable credential/server snapshot after env consumption. `newFSClientForContext()` calls `loadConfig()` every time an explicit context path is resolved. A single command can therefore combine:

- default/current client built from one resolver snapshot, and
- explicit-context client built from a later disk read.

If another terminal runs `drive9 ctx use` or edits config between calls, one command can operate with mixed config views. That is exactly the class of routing inconsistency this PR is trying to fix.

Required fix: pass a shared `*Config`/resolver snapshot into explicit-context client creation, or otherwise make config reads consistent for the command invocation. Add a regression test that proves explicit-context routing does not observe a later config mutation within one command path.

## Non-blocking cleanup

- Several command handlers discard unused return values from `fsClientForRemoteArg`; this makes the helper API noisier than needed. Consider returning a smaller shape or using a small struct.
- `cp.go` now has repeated `if Context != "" { newFSClientForContext(...) }` blocks. Not blocking, but a small helper would reduce future divergence between recursive and non-recursive paths.

I’ll re-review after the server precedence decision is made and implemented consistently.

**Comment by srstack** [comment]
Resolved the latest review blockers:

- B1: Kept the intended PR behavior: context server takes precedence over `DRIVE9_SERVER`. Updated `docs/specs/vault-interaction-end-state.md` and `docs/guides/vault-quickstart.md` to document the new precedence, explicit `<ctx>:/path` fallback behavior, and migration guidance for CI / one-off usage.
- B2: `newFSClientForContext()` now falls back to `DRIVE9_SERVER -> config.server -> default` when the named context has no `server`, so the explicit-context path no longer skips the env server fallback.
- B3: `ResolveCredentials()` and explicit-context FS client creation now share the same per-process config snapshot, preventing a single command from observing different config versions for the default client and explicit-context client.

Added regression coverage for:

- Explicit context using `DRIVE9_SERVER` fallback.
- Explicit context not observing config mutations after the resolver snapshot is taken.

Verified with:

- `go test ./cmd/drive9/cli -count=1`
- `go build -o bin/drive9 ./cmd/drive9`

Pushed in `6534aa2`.

**Review by qiffang — COMMENTED** [review]
## adversary-2 Round 2 review — PR #454 (head 6534aa21)

Round 1 blockers resolved:

**B3** (config re-read TOCTOU) ✅ — `newFSClientForContext` now uses `loadCredentialConfigSnapshot()` which is `sync.Once`-cached. Same config snapshot as `ResolveCredentials`. No more double disk reads.

**B4** (DRIVE9_SERVER skipped in explicit-context path) ✅ — `newFSClientForContext` reads `r.envServer` from the cached `ResolvedCredentials`, which holds the consumed env value. Explicit-context server fallback is now: `ctx.Server > r.envServer > cfg.Server > defaultServerURL` — consistent with the unified resolver.

**Spec docs updated** ✅ — `vault-quickstart.md` and `vault-interaction-end-state.md` both updated with the new precedence order and migration notes for CI/container users who relied on `DRIVE9_SERVER` overriding active context server.

**New tests** ✅ — `TestExplicitContextUsesEnvServerFallback` covers the DRIVE9_SERVER fallback for explicit contexts. `TestExplicitContextUsesResolverConfigSnapshot` verifies config mutation after snapshot doesn't affect behavior.

**Round 1 warnings status:**
- W1 (mv.go dead variables): Still present, not blocking.
- W2 (no defaultServerURL fallback test): Not explicitly added, but covered implicitly by existing tests.
- W3 (unused return values): Still present, not blocking.
- W4 (cp.go duplication): Still present, not blocking — style issue.

No new issues found.

**Verdict: GREEN** ✅

**Review by qiffang — APPROVED** [review]

**Review by qiffang — APPROVED** [review]
## dat9-dev1 Round 2 — PR #454 (head `6534aa2`) — **GREEN** ✅

### Round 1 blockers — all resolved

**B1** (server precedence spec contradiction) ✅ — Intentional change, now `ctx.Server > DRIVE9_SERVER > cfg.Server > default`. Both spec docs (vault-quickstart.md §14.2, vault-interaction-end-state.md §14.2) updated with new precedence + migration note for CI jobs that relied on old DRIVE9_SERVER-wins behavior.

**B2** (DRIVE9_SERVER fallback missing in `newFSClientForContext`) ✅ — `newFSClientForContext` now reads `r.envServer` from the consumed env var, fallback chain is `ctx.Server > envServer > cfg.Server > defaultServerURL`. This correctly avoids re-reading the unset env var.

### New commit analysis (`6534aa2`)

**Config snapshot consistency** ✅ — `loadCredentialConfigSnapshot()` with `sync.Once` ensures both `ResolveCredentials()` and `newFSClientForContext()` share the same disk-read config. Eliminates the TOCTOU window where config could change between resolver and explicit-context client creation. `resetCredentialCacheForTest()` properly resets both `sync.Once` guards.

**Cross-context guards** ✅ — `cp` (both recursive and non-recursive) and `mv` reject mixed explicit/current-context operations with clear error messages. The switch pattern `case srcRP.Context == "" && dstRP.Context == "":` correctly handles the "both current context" case as a no-op.

**Test coverage** ✅ — 7 new tests in `fs_context_test.go` covering:
- Explicit remote context routing (`TestLsUsesExplicitRemoteContext`)
- Delegated context rejection (`TestLsRejectsDelegatedExplicitRemoteContext`)
- Config server fallback for explicit context (`TestExplicitContextUsesConfigServerFallbackInsteadOfActiveContextServer`)
- Env server fallback when named context has no server (`TestExplicitContextUsesEnvServerFallback`)
- Config snapshot consistency (`TestExplicitContextUsesResolverConfigSnapshot`)
- Cross-context cp/mv rejection (3 tests)

**Credential tests** ✅ — `TestResolveCredentials_ContextServerBeatsEnvServer` now asserts the new precedence. Two new tests cover the cross-axis case: context server wins even when credential comes from env API key or env vault token.

### Minor observation (non-blocking)

`envServer` is stored as an unexported field on `ResolvedCredentials`. This works because `newFSClientForContext` calls `ResolveCredentials()` to access it. The field is effectively internal state rather than part of the public contract — which is correct, since callers should not depend on it. No change needed.

Ship it.

**Review by qiffang — APPROVED** [review]
Round 2 review: GREEN.

I re-reviewed head `6534aa2` against the prior blockers and the latest diff. The earlier server-resolution issues are resolved:

- `ResolveCredentials()` now uses one cached config snapshot via `loadCredentialConfigSnapshot()`, and `newFSClientForContext()` reads the same snapshot, so one command no longer mixes resolver credentials from one config read with explicit-context FS clients from another config read.
- Explicit `<ctx>:/path` routing now falls back through the consumed `DRIVE9_SERVER` value (`ctx.Server > envServer > cfg.Server > default`) instead of skipping the env server.
- The server precedence change is now documented in both `vault-quickstart.md` and `vault-interaction-end-state.md`, including a migration note for users who relied on `DRIVE9_SERVER` overriding an active context server.
- Cross-context copy/rename is rejected instead of silently using the wrong client.

Validation I ran locally:

- `/usr/local/go/bin/go test ./cmd/drive9/cli -count=1`
- `/usr/local/go/bin/go test ./cmd/drive9/cli -run 'ResolveCredentials|ExplicitContext|FsContext|RemoteContext|Ctx|Cp|Mv|Mount' -count=1`
- `/usr/local/go/bin/go build -o /tmp/drive9-pr454-drive9 ./cmd/drive9`
- `git diff --check origin/main...HEAD`

Residual non-blocking note: the tests cover mixed current/explicit cross-context rejection. I would still add a future regression for two different explicit contexts (`prod:/a` -> `staging:/b`) because the code rejects that path too, but this is not blocking the current fix.

**Review by qiffang — APPROVED** [review]
PR #454 Round 2 re-review — APPROVE

I re-reviewed latest head `6534aa21fdd7cbae92d9745bb57c7d0df32957fc` against the PR API base `e58fc8ee41588023edddaed1c3fe28a229845032`.

Previous blockers are resolved:

- Server precedence is now explicit and consistent: active context `server` > `DRIVE9_SERVER` > top-level config `server` > compiled default. Docs/spec were updated with the changed precedence and migration note.
- Explicit filesystem contexts no longer skip `DRIVE9_SERVER`: `newFSClientForContext` uses named `ctx.Server`, then the resolver-captured `envServer`, then config/default. It also does not incorrectly inherit the active context server for `<ctx>:/path`.
- Config TOCTOU is fixed: `loadCredentialConfigSnapshot()` uses `sync.Once`, and both `ResolveCredentials` and `newFSClientForContext` read the same per-process config snapshot.
- Regression coverage now includes explicit-context env fallback, resolver snapshot consistency, context-vs-env precedence, and cross-context cp/mv rejection.

Validation run locally:

- `go test ./cmd/drive9/cli -count=1`
- `go test ./cmd/drive9/cli -run 'ResolveCredentials|ExplicitContext|RemoteArg|Ctx|Mount|Vault' -count=1`
- `go build -o /tmp/drive9-pr454-drive9 ./cmd/drive9`
- `git diff --check e58fc8ee41588023edddaed1c3fe28a229845032..6534aa21fdd7cbae92d9745bb57c7d0df32957fc`

GitHub checks are also green (`ci`, `drive9-server-local e2e smoke`; CodeRabbit pass/skipped). No blockers from my side.
